src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -28476,6 +28476,263 @@ void M2()
+public class C
+{
+    [return: NotNullIfNotNull(""c"")]
+    public static explicit operator string?(C? c) => c?.ToString();
+
+    void M1(string s) { }
+
+    void M2()
+    {
+        M1((string)new C());
+        M1((string?)(C?)null); // 1
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { NotNullIfNotNullAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (13,12): warning CS8604: Possible null reference argument for parameter 's' in 'void C.M1(string s)'.
+                //         M1((string?)(C?)null); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "(string?)(C?)null").WithArguments("s", "void C.M1(string s)").WithLocation(13, 12));
+        }
+
+        [Fact]
+        [WorkItem(39802, "https://github.com/dotnet/roslyn/issues/39802")]
+        public void UserDefinedConversion_UnconditionalParamAttributes()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public struct A { }
+public struct B
+{
+    public static implicit operator A([DisallowNull] B? b)
+    {
+        b.Value.ToString();
+        return new A();
+    }
+
+    void M1(A a) { }
+
+    void M2()
+    {
+        M1(new B());
+        M1((B?)null); // 1
+    }
+}
+
+public class C
+{
+    public static implicit operator string?([AllowNull] C c) => c?.ToString();
+
+    void M1(string? s) { }
+    void M2()
+    {
+        M1(new C());
+        M1((C?)null);
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { AllowNullAttributeDefinition, DisallowNullAttributeDefinition, source });
+            comp.VerifyDiagnostics(
+                // (18,12): warning CS8607: A possible null value may not be used for a type marked with [NotNull] or [DisallowNull]
+                //         M1((B?)null); // 1
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "(B?)null").WithLocation(18, 12));
+        }
+
+        [Fact]
+        [WorkItem(39802, "https://github.com/dotnet/roslyn/issues/39802")]
+        public void UserDefinedConversion_UnconditionalParamAttributes_Lifted()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+
+public struct A { }
+
+public struct B
+{
+    public static implicit operator A([DisallowNull] B b)
+    {
+        return new A();
+    }
+
+    void M0(A? a) { }
+    void M1([DisallowNull] A? a) { }
+
+    void M2()
+    {
+        M0((B?)new B());
+        M0((B?)null);
+
+        M1((B?)new B());
+        M1((B?)null); // 1
+    }
+}
+
+public struct C
+{
+    [return: NotNull]
+    public static implicit operator A?([DisallowNull] C c)
+    {
+        return new A();
+    }
+
+    void M0(A? a) { }
+    void M1([DisallowNull] A? a) { }
+
+    void M2()
+    {
+        M0((C?)new C());
+        M0((C?)null);
+
+        M1((C?)new C());
+        M1((C?)null); // 2
+    }
+}
+
+public struct D
+{
+    [return: NotNullIfNotNull(""d"")]
```

Comment:
>NotNullIfNotNull(""d"") [](start = 13, length = 23)

`d` is never `null`.

Reply:
I wanted to verify that this postcondition is respected, even in this contrived scenario, but it feels maybe interesting for the input to be System.Nullable for this and for `operator A?(C c)` above.

Reply:
I think that although this is not a reasonable real-world coding pattern (user should just use [NotNull] instead), it is good to record the behavior of this scenario. I may get rid of `[DisallowNull]`, though.